src/import-export/csv-imp/gnc-tx-import.cpp

```diff
@@ -0,0 +1,960 @@
+    if (orig_lines.size() == 0)
+    {
+        g_set_error (error, GNC_CSV_IMP_ERROR, GNC_CSV_IMP_ERROR_PARSE, "Parsing failed.");
+        return 1;
+    }
+
+    if (guessColTypes)
+    {
+        /* Free column_types if it's already been created. */
+        column_types.clear();
+    }
+    column_types.resize(max_cols, GncTransPropType::NONE);
+
+    if (guessColTypes)
+    {
+        /* Guess column_types based
+         * on the contents of each column. */
+        /* TODO Make it actually guess. */
+    }
+    return 0;
+}
+
+
+/** Convert str into a time64 using the user-specified (import) date format.
+ * @param str The string to be parsed
+ * @param date_format The date format to use.
+ * @return a pointer to a time64 on success, nullptr on failure
+ */
+static time64* convert_date_col_str (const std::string &str, int date_format)
+{
+    auto parsed_date = parse_date (str.c_str(), date_format);
+    if (parsed_date == -1)
+        return nullptr;
+    else
+    {
+        auto mydate = new time64;
+        *mydate = parsed_date;
+        return mydate;
+    }
+}
+
+
+/** Convert str into a gnc_numeric using the user-specified (import) currency format.
+ * @param str The string to be parsed
+ * @param currency_format The currency format to use.
+ * @return a pointer to a gnc_numeric on success, nullptr on failure
+ */
+static gnc_numeric* convert_amount_col_str (const std::string &str, int currency_format)
+{
+    /* If a cell is empty or just spaces return 0 as amount */
+    if(!boost::regex_search(str, boost::regex("[0-9]")))
+        return nullptr;
+
+    auto expr = boost::make_u32regex("[[:Sc:]]");
+    std::string str_no_symbols = boost::u32regex_replace(str, expr, "");
+
+    /* Convert based on user chosen currency format */
+    gnc_numeric val;
+    char *endptr;
+    switch (currency_format)
+    {
+    case 0:
+        /* Currency locale */
+        if (!(xaccParseAmount (str_no_symbols.c_str(), TRUE, &val, &endptr)))
+            return nullptr;
+        break;
+    case 1:
+        /* Currency decimal period */
+        if (!(xaccParseAmountExtended (str_no_symbols.c_str(), TRUE, '-', '.', ',', "\003\003", "$+", &val, &endptr)))
+            return nullptr;
+        break;
+    case 2:
+        /* Currency decimal comma */
+        if (!(xaccParseAmountExtended (str_no_symbols.c_str(), TRUE, '-', ',', '.', "\003\003", "$+", &val, &endptr)))
+            return nullptr;
+        break;
+    }
+
+    auto amount = new gnc_numeric;
+    *amount = val;
+    return amount;
+}
+
+/* Define a class hierarchy to temporarily store transaction/split properties
+ * found in one import line. There is a generic parent class and an implementation
+ * template class. This template class is further specialized for each data type
+ * we support (currently time64, Account, string and gnc_numeric).
+ */
+struct GncTransProperty
+{
+    virtual ~GncTransProperty()
+      //Remove pure designation.
```

Comment:
Well, I understood these would have to exist already. I was playing with the idea of creating those as soon as the user selects the first column that's transaction related. The same for splits. When the first column for a split related type is selected, create a split for each row. If I do this via a unique pointer with a custom deleter that destroys the transaction/split if it goes out of scope, I don't have to worry about memory management. When all available transaction and split data is set, I can release the known good tx's and splits from the unique pointer and pass them on to the generic importer (which currently expects a GList of transactions that can still be incomplete). I believe this can work.

The tricky part is still in creating the necessary GncTransProperty objects. Most of them are easy as they all map on to one to either a transaction or a split property (like date, description,...). The only difficult ones are amount (which can be deposit or withdrawal but that's unimportant in this discussion) and balance.

Amount is tricky because if the user selects an "Other account", two splits will be created. One having amount as value and the other -amount as value. So in this case amount should be used twice. If the user first selects an amount only one GncTransAmountProperty is created, with only the first split as parameter. So iterating over the property vector later to fill in transaction and splits will not find a property to set the amount of the second split. I could add special code at user column selection time that checks
1. when amount is selected whether an other account property exists already and then adds two amount properties, one for the first split and one for the second.
1. when other account is selected whether an amount property exists already and then adds another one for the second split with negative amount.

Special code is then also needed to drop these properties again if the user changes his mind and unsets the other account column...

A balance column is even weirder. In that case the amounts are calculated based on the balance of the previous transaction and the current one. This is not mapped easily in the set_prop function, because the current implementation also takes already existing transactions into account. I wonder who ever invented this option...